compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1998,8 +2000,24 @@ public void visit(BLangInvocation invocationExpr) {
         }
 
         if (invocationExpr.actionInvocation || invocationExpr.async) {
-            validateActionInvocation(invocationExpr.pos, invocationExpr);
+            if (invocationExpr.async && this.withinLockBlock) {
+                logAsyncCallWithinLockError(invocationExpr);
+            } else {
+                validateActionInvocation(invocationExpr.pos, invocationExpr);
+            }
+        }
+    }
+
+    private void logAsyncCallWithinLockError(BLangInvocation invocationExpr) {
+        DiagnosticCode code;
+
+        if (invocationExpr.functionPointerInvocation) {
+            code = DiagnosticCode.USAGE_OF_WORKER_WITHIN_LOCK_IS_PROHIBITED;
+        } else {
+            code = DiagnosticCode.USAGE_OF_START_WITHIN_LOCK_IS_PROHIBITED;
         }
+
+        dlog.error(invocationExpr.pos, code);
```

Comment:
We can simplify this using a ternary operator?
```suggestion
        dlog.error(invocationExpr.pos, invocationExpr.functionPointerInvocation ?  
                   DiagnosticCode.USAGE_OF_WORKER_WITHIN_LOCK_IS_PROHIBITED : 
                   DiagnosticCode.USAGE_OF_START_WITHIN_LOCK_IS_PROHIBITED);
```

Reply:
Fixed